File: compilers/go/frontend/diagnostic.go
package frontend

import "fmt"

// Severity represents the level of a compiler diagnostic.
type Severity string

const (
	SeverityError   Severity = "error"
	SeverityWarning Severity = "warning"
)

// Diagnostic represents a single compiler diagnostic (error or warning).
type Diagnostic struct {
	Message  string
	Loc      *SourceLocation
	Severity Severity
}

// MakeDiagnostic creates a diagnostic with the given message, severity, and optional location.
func MakeDiagnostic(message string, severity Severity, loc *SourceLocation) Diagnostic {
	return Diagnostic{Message: message, Severity: severity, Loc: loc}
}

// FormatMessage returns the diagnostic message with optional file:line:column prefix.
func (d Diagnostic) FormatMessage() string {
	if d.Loc != nil && d.Loc.File != "" {
		if d.Loc.Column > 0 {
			return fmt.Sprintf("%s:%d:%d: %s", d.Loc.File, d.Loc.Line, d.Loc.Column, d.Message)
		}
		return fmt.Sprintf("%s:%d: %s", d.Loc.File, d.Loc.Line, d.Message)
	}
	return d.Message
}

File: compilers/ruby/lib/runar_compiler/frontend/diagnostic.rb
# frozen_string_literal: true

# Structured compiler diagnostics.
#
# This module defines the Diagnostic class used to report errors and warnings
# during compilation. It is a direct port of
# compilers/python/runar_compiler/frontend/diagnostic.py.

require_relative "ast_nodes"

module RunarCompiler
  module Frontend
    # Severity constants for diagnostics.
    module Severity
      ERROR   = "error"
      WARNING = "warning"
    end

    # A single compiler diagnostic (error or warning).
    Diagnostic = Struct.new(:message, :severity, :loc, keyword_init: true) do
      def initialize(message:, severity:, loc: nil)
        super
      end

      # Format with optional file:line:column prefix.
      def format_message
        if loc && !loc.file.empty?
          if loc.column > 0
            return "#{loc.file}:#{loc.line}:#{loc.column}: #{message}"
          end

          return "#{loc.file}:#{loc.line}: #{message}"
        end

        message
      end

      def to_s
        format_message
      end
    end
  end
end
